zenmap: init at 7.95

mymindstorm opened the pull request:
Fixes #287288

Zenmap is the offical GUI for nmap, a network scanner. It was previously included in nixpkgs, but disabled due to requiring python 2. Since then, upstream has migrated to python 3.

Zenmap is also a bit strange to package because its source code lives in a subdirectory of nmap's source code. A new package seemed most appropriate because I felt that adding an option to nmap would overcomplicate things. I was also unable to build it using nmap's makefile, whereas `buildPythonPackage` works perfectly. 

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

mymindstorm committed:
maintainers: add mymindstorm

mymindstorm committed:
zenmap: init at 7.95 #287288

ethancedwards8 commented on pkgs/by-name/ze/zenmap/package.nix:
```diff
@@ -0,0 +1,46 @@
+    setuptools
+  ];
+
+  nativeBuildInputs = [
+    wrapGAppsHook3
+    gobject-introspection
+  ];
```
may could just go in deps? not sure.

phaer replied:
If it's a library we link against and need at runtime (not only build-time) it should go into `buildInputs`.

ethancedwards8 reviewed (approved):
Diff LGTM.

awwpotato commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 393262`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>zenmap</li>
    <li>zenmap.dist</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>zenmap</li>
    <li>zenmap.dist</li>
  </ul>
</details>

awwpotato commented:
> @ethancedwards8 @awwpotato is there any chance that this pr could get merged? I'm new to nix, so I am unsure as to the process to request a merge.

I can't merge this, but I'll ask for someone on matrix to take a look (and hopefully merge).

Henry-Hiles commented:
wait why was maintainer list removed

Henry-Hiles commented:
Sorry about that, I think its a bug with github on mobile.

phaer commented on pkgs/by-name/ze/zenmap/package.nix:
```diff
@@ -0,0 +1,46 @@
+
+  dependencies = [
+    python3Packages.pygobject3
+    nmap
```
This isn't a python package, but normal `nmap`, right? If so, it should probably go into `buildInputs`. `dependencies` is for python things.

phaer reviewed (commented):
Two small remarks regarding dependencies, otherwise LGTM. Happy to merge once those are resolved! :)

dvaerum commented:
Hey @mymindstorm,

You can add me to the maintainer list if you like, I just made a pull request for packaging Zenmap.
https://github.com/NixOS/nixpkgs/pull/401785
However, I have closed it again since you are already working on it.

Additionally, there may be some things you can use from my packages because it does not look like you have included the desktop-files. Anyway, feel free to copy-paste if there is anything you can use.
I got Zenmap working both there it is running as root and as a normal user.

FliegendeWurst commented on pkgs/by-name/ze/zenmap/package.nix:
```diff
@@ -0,0 +1,46 @@
+  gobject-introspection,
+}:
+
+python3Packages.buildPythonPackage rec {
```
```suggestion
python3Packages.buildPythonApplication rec {
```
Is recommended if you package an application

FliegendeWurst commented on pkgs/by-name/ze/zenmap/package.nix:
```diff
@@ -0,0 +1,46 @@
+
+  meta = {
+    description = "Offical nmap Security Scanner GUI";
+    homepage = "http://www.nmap.org";
```
```suggestion
    homepage = "https://nmap.org/zenmap/";
```

mymindstorm commented:
Thanks for the reviews!

@dvaerum I've incorporated parts of your PR to this one and tested the elevate to root script successfully

@phaer @FliegendeWurst I think I've addressed your review comments.

FliegendeWurst commented on pkgs/by-name/ze/zenmap/package.nix:
```diff
@@ -0,0 +1,76 @@
+
+  src = nmap.src;
+
+  sourceRoot = "${nmap.name}/zenmap";
```
```suggestion
  prePatch = ''
    cd zenmap
  '';
```
`nmap.name` is not always nmap-7.95 (e.g. for cross or other configurations)

FliegendeWurst commented on pkgs/by-name/ze/zenmap/package.nix:
```diff
@@ -0,0 +1,76 @@
+    install -Dm 644 "zenmapCore/data/pixmaps/zenmap.png" -t "$out/share/pixmaps/"
+    # Desktop-files for application
+    install -Dm 644 "install_scripts/unix/zenmap.desktop" -t "$out/share/applications/"
+    substituteInPlace "$out/share/applications/zenmap.desktop" \
+        --replace-fail "Exec=zenmap %F" \
+                      "Exec=$out/bin/zenmap %F"
+    install -Dm 644 "install_scripts/unix/zenmap-root.desktop" -t "$out/share/applications/"
+    substituteInPlace "$out/share/applications/zenmap-root.desktop" \
+        --replace-fail "TryExec=su-to-zenmap.sh" \
+                      "TryExec=$out/bin/su-to-zenmap.sh" \
+        --replace-fail "Exec=su-to-zenmap.sh %F" \
+                      "Exec=$out/bin/su-to-zenmap.sh %F"
```
```suggestion
    install -Dm 644 "install_scripts/unix/zenmap-root.desktop" -t "$out/share/applications/"
```
#308324

FliegendeWurst commented on pkgs/by-name/ze/zenmap/package.nix:
```diff
@@ -0,0 +1,76 @@
+                      '"${xterm}/bin/xterm"'
+  '';
+
+  checkPhase = ''
+    cd test
+    ${python3Packages.python.interpreter} run_tests.py
+  '';
```
```suggestion
  checkPhase = ''
    runHook preCheck

    cd test
    ${python3Packages.python.interpreter} run_tests.py 2>&1 | tee /dev/stderr | tail -n1 | grep '^OK$'

    runHook postCheck
  '';
```
The check script always exits with code 0, so we must check for passing tests.
When overriding a phase the pre and post hooks must be run.

FliegendeWurst commented on pkgs/by-name/ze/zenmap/package.nix:
```diff
@@ -0,0 +1,76 @@
+  dependencies = with python3Packages; [
+    pygobject3
+  ];
+
```
```suggestion
  nativeCheckInputs = [
    nmap
  ];
```
Required by the test suite

mymindstorm commented:
@FliegendeWurst thanks for the review! The most recent version should address your comments.

FliegendeWurst commented on pkgs/by-name/ze/zenmap/package.nix:
```diff
@@ -0,0 +1,78 @@
+    pygobject3
+  ];
+
+  postInstall = ''
+    wrapProgram "$out/bin/zenmap" \
+      --prefix PATH : ${lib.makeBinPath [ nmap ]}
```
```suggestion
  dontWrapGApps = true;

  postInstall = ''
    makeWrapperArgs+=("''${gappsWrapperArgs[@]}")
    makeWrapperArgs+=(--prefix PATH : ${lib.makeBinPath [ nmap ]})
```
Otherwise you get this mess of wrappers :wink:  (Python + GApps + your PATH wrapper)
```
arne@nixOS> \ls -lah /nix/store/7sr0cnai7pwxcrnkxhppaw1x64iaasf9-zenmap-7.95/bin                                                                                                                                              ~/nixpkgs-wt-2
total 100K
dr-xr-xr-x 1 root root  266 1970-01-01 01:00 .
dr-xr-xr-x 1 root root   44 1970-01-01 01:00 ..
-r-xr-xr-x 1 root root  20K 1970-01-01 01:00 su-to-zenmap.sh
-r-xr-xr-x 1 root root 1,9K 1970-01-01 01:00 .su-to-zenmap.sh-wrapped
-r-xr-xr-x 1 root root  20K 1970-01-01 01:00 zenmap
-r-xr-xr-x 1 root root  20K 1970-01-01 01:00 .zenmap-wrapped
-r-xr-xr-x 1 root root  16K 1970-01-01 01:00 .zenmap-wrapped_
-r-xr-xr-x 1 root root  16K 1970-01-01 01:00 ..zenmap-wrapped-wrapped
-r-xr-xr-x 1 root root  780 1970-01-01 01:00 ...zenmap-wrapped-wrapped-wrapped
```
